Feature/MountedViewsProvider

JonnyDawe opened the pull request:
reference: #11 

# Description
This pull request introduces a new MountedViewsProvider component, allowing users to seamlessly access Views outside of the component that directly renders the map. The MapProvider wraps all nodes where map access is desired.

# Changes Made
- Added a new MountedViewsProvider component.
- The MountedViewsProvider component can be used to wrap nodes where map access is needed.
- The implementation follows a pattern similar to the example in [visgl/react-map-gl](https://visgl.github.io/react-map-gl/docs/api-reference/map-provider), with adaptations to match the specific patterns of this project.
- Added a example.

Very happy for changes and feedback!

am-maneaters commented:
this is looking awesome, i'm gonna take some time to review it today

am-maneaters commented on apps/web/src/examples/MountedViews.tsx:
```diff
@@ -0,0 +1,94 @@
+  );
+}
+
+function BaseMapPickList() {
+  const views = useMountedViews();
+  if (!views) return null;
+
+  return (
+    <CalciteBlock heading="Basemap options" open>
+      {Object.entries(views).map(([id, view]) => {
+        if (!view) return null;
+
+        return (
+          <BasemapPicker key={id} label={`Set Basemap for ${id}`} view={view} />
+        );
+      })}
+    </CalciteBlock>
+  );
+}
```
```suggestion
function BaseMapPickList() {
  const { mySceneView, myMapView } = useMountedViews();

  return (
    <CalciteBlock heading="Basemap options" open>
      {myMapView && (
        <BasemapPicker label="Set Basemap for mapView" view={myMapView} />
      )}
      {mySceneView && (
        <BasemapPicker label="Set Basemap for sceneView" view={mySceneView} />
      )}
      <ArcBasemapGallery view={mySceneView} />
    </CalciteBlock>
  );
}
```

i feel like this might be a better way to convey to the users that you can get a view directly from their ID that you define on the ArcMapView/ArcSceneView.

am-maneaters commented on apps/web/src/examples/MountedViews.tsx:
```diff
@@ -0,0 +1,94 @@
+  );
+}
+
+function BasemapPicker({
+  view,
+  label,
+}: {
+  view: __esri.MapView | __esri.SceneView;
+  label: string;
+}) {
+  const [basemap, setBasemap] = useArcState(view.map, 'basemap');
+  console.log(basemap.id);
+
+  return (
+    <CalciteLabel>
+      {label}
+      <CalciteSelect
+        label="Basemap"
+        value={basemap.id}
+        onCalciteSelectChange={(e) =>
+          setBasemap(Basemap.fromId(e.target.value))
+        }
+      >
+        <CalciteOption value="streets">Streets</CalciteOption>
+        <CalciteOption value="terrain">Terrain</CalciteOption>
+        <CalciteOption value="dark-gray-vector">Dark Gray Vector</CalciteOption>
+      </CalciteSelect>
+    </CalciteLabel>
+  );
+}
```
could we just use the ArcBasemapGallery instead of this custom component? just so we can let the user focus more on the rest of the example

am-maneaters commented on packages/arcgis-react/components/ArcView/MountedViewsContext.tsx:
```diff
@@ -0,0 +1,81 @@
+>(undefined);
+
+export const MountedViewsProvider = ({ children }: React.PropsWithChildren) => {
+  const [view, setViews] = useState<{ [id: string]: MapView | SceneView }>({});
```
```suggestion
  const [views, setViews] = useState<{ [id: string]: MapView | SceneView }>({});
```

am-maneaters commented on packages/arcgis-react/components/ArcView/MountedViewsContext.tsx:
```diff
@@ -0,0 +1,81 @@
+export const MountedViewsProvider = ({ children }: React.PropsWithChildren) => {
+  const [view, setViews] = useState<{ [id: string]: MapView | SceneView }>({});
+
+  const onMapMount = useCallback((view: MapView | SceneView, id: string) => {
```
```suggestion
  const onViewMount = useCallback((view: MapView | SceneView, id: string) => {
```

JonnyDawe replied:
Cleaned up per suggestion

am-maneaters commented on packages/arcgis-react/components/ArcView/MountedViewsContext.tsx:
```diff
@@ -0,0 +1,81 @@
+    });
+  }, []);
+
+  const onMapUnmount = useCallback((id: string) => {
```
```suggestion
  const onViewUnmount = useCallback((id: string) => {
```

am-maneaters commented on packages/arcgis-react/components/ArcView/MountedViewsContext.tsx:
```diff
@@ -0,0 +1,81 @@
+  return (
+    <MountedViewsContext.Provider
+      value={{
+        views: view,
+        onViewMount: onMapMount,
+        onViewUnmount: onMapUnmount,
```
```suggestion
        views,
        onViewMount,
        onViewUnmount,
```

am-maneaters commented on packages/arcgis-react/components/util/createView.tsx:
```diff
@@ -39,24 +48,36 @@ export function createViewComponent<
 
     const mapContainer = useRef<HTMLDivElement>(null);
 
+    const internalId = useId();
```
remove destructured id property and use:
```suggestion
    const internalId = useId();
    const id = mapViewProps.id ?? internalId;
```

then refactor all instances of `id ?? internalId` to `id`

am-maneaters commented:
@JonnyDawe, do you think it would make sense to merge the `useView` and `useMountedViews` into a single hook? i'm looking at the `useMap` hook for [React Map GL](https://visgl.github.io/react-map-gl/docs/api-reference/use-map), and i like how the hook can either be used as `const { current: map } = useMap()` and also `const { myFirstMap, mySecondMap } = useMap()`. 

so basically, instead of even adding the `useMountedViews` hook, could we just update the existing `useView` hook?

JonnyDawe commented:
@am-maneaters, I agree with all of your suggestions above, including the changes to the example to use `ArcBasemapGallery` (once the details of whether to merge the hooks logic is worked out!)

I agree that it would make sense to consolidate the logic to have just a single hook for working with the views. The added benefit being of removing my clunky naming of `useMountedViews`!

If we were to merge the logic into useView, I can see that the hooks useMapView / useSceneView conceptually might not make sense anymore as we won't necessarily be working with a single view? 

We would also need to refactor createLayer, createWidget and useArcUI to use `const { current: view} = useView()`. Are you happy with that change?

JonnyDawe commented:
I have changed useMountedViews to now be useView. At the same time I made change to the existing `useView` , `useMapView` and `useSceneView` to be `useCurrentView`. This hook works with the new `useView` hook and `useMountedViews` provider. 

The example has now also been simplified and updated. 

@am-maneaters - do these changes fit with your existing patterns? I was wondering whether we should give the `MountedViewsProvider` a cleaner name, e.g. just `ViewsProvider`

JonnyDawe commented:
@am-maneaters - I've pulled a lot of this logic into another project I'm working on and it seems to be holding up nicely! 

Having the map views available in a provider has made it really straightforward to move around the DOM structure - e.g. placing a custom search component (radix/searchVM) into a header component outside of the Map hierarchy.

am-maneaters reviewed (approved):
this is looking great, i'm good to merge if you are!

JonnyDawe commented:
@am-maneaters - good shout, I'm in agreement!

JonnyDawe commented:
@am-maneaters - should be updated now :)
